src/preact/base-element.js

```diff
@@ -265,20 +273,43 @@ export class PreactBaseElement extends AMP.BaseElement {
             'when configured with "children", "passthrough", or ' +
             '"passthroughNonEmpty" properties.'
         );
-        const shadowRoot = this.element.attachShadow({mode: 'open'});
-        this.container_ = shadowRoot;
-
-        const shadowCss = Ctor['shadowCss'];
-        if (shadowCss) {
-          installShadowStyle(shadowRoot, this.element.tagName, shadowCss);
+        // Check if there's a pre-constructed shadow DOM.
+        let {shadowRoot} = this.element;
+        let container = shadowRoot && childElementByTag(shadowRoot, 'c');
```

Comment:
Do we really need to create a new `<c>` element? Couldn't we just infer that an already created shadow root means that this was SSR'd?

Reply:
I didn't find a stable enough way tbh. Here's what happens here. The shadow root looks like this

```
#shadow-root
  <style>...</style>
  <slot name="i-amphtml-svc"></slot>
  <c>
     <component-root>...</component-root>
  </c>
```

The `component-root` is whatever element that the hydration will go into. The `hydrate` API has to be called on the container. E.g.

```
hydrate(c, Component)
```

If I were to call `hydrate()` on a shadow root, it can get confused. E.g. if the root of a component was a `slot` element (e.g. in the passthrough mode), it'd think that `<slot name="i-amphtml-svc">` is the one to hydrate with.

So this is a very unstable situation. It's less likely to collide on `<style>`, of course. But any other children of shadow root would be up for grabs. Unfortunately, there's no API to say "hydrate only this node" - it only takes the whole container.

Reply:
Thank you for the thorough explanation! I'm surprised that Preact gets confused about which node to hydrate given multiple options

Reply:
The first render/hydrate is most important when the framework is mounted. So I don't think it's that surprising give the current APIs. I think it'd have been relatively simple thing if APIs were different. I'll keep playing around to see if we can find a better hydration approach. But it's a tradeoff between trying to rely on some undocumented nuances of the implementation vs what API says.